docs/core/diagnostics/metrics-collection.md

```diff
@@ -0,0 +1,404 @@
+[hats-sold-graph](media\prometheus-hat-sold-metric.png)
+
+If the Prometheus server hasn't been scraping (Prometheus' terminology for polling) the sample app for long you may need to wait a short
+while for data to accumulate. The default configuration has Prometheus scrape metric data every 15 seconds.
+
+#### Showing metrics on a Grafana dashboard
+
+1. Follow [the standard instructions](https://prometheus.io/docs/visualization/grafana/#creating-a-prometheus-graph) to install Grafana and
+connect it to a Prometheus data source.
+
+2. Create a Grafana dashboard by clicking the + icon on the left toolbar in the Grafana web portal, then select 'Dashboard'. In dashboard
+editor that appears enter 'Hats Sold/Sec' as the Title and 'rate(hats_sold[5m])' in the PromQL expression field. It should look like this:
+[hats-sold-dashboard-editor](media\grafana-hats-sold-dashboard-editor.png)
+
+3. Click Apply to save and view our simple new dashboard.
+[hats-sold-dashboard](media\grafana-hats-sold-dashboard)
+
+
+### Creating a custom collection tool using the .NET <xref:System.Diagnostics.Metrics.MeterListener> API
+
+The .NET <xref:System.Diagnostics.Metrics.MeterListener> API allows creating custom in-process logic to observe the measurements
+being recorded by <xref:System.Diagnostics.Metrics.Meter?displayProperty=nameWithType>. For guidance creating custom
+logic compatible with the older EventCounters instrumentation see [here](event-counters.md).
+
+Modify the code of Program.cs to use the MeterListener like this:
+```C#
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Metrics;
+using System.Threading;
+
+class Program
+{
+    static Meter s_meter = new Meter("HatCo.HatStore", "1.0.0");
+    static Counter<int> s_hatsSold = s_meter.CreateCounter<int>(name: "hats-sold", 
+                                                                unit: "Hats",
+                                                                description: "The number of hats sold in our store"); 
+
+    static void Main(string[] args)
+    {
+        MeterListener meterListener = new MeterListener();
+        meterListener.InstrumentPublished = (instrument, listener) =>
+        {
+            if(instrument.Meter.Name == "HatCo.HatStore")
+            {
+                listener.EnableMeasurementEvents(instrument);
+            }
+        };
+        meterListener.SetMeasurementEventCallback<int>(OnMeasurementRecorded);
+        meterListener.Start();
+
+        Console.WriteLine("Press any key to exit");
+        while(!Console.KeyAvailable)
+        {
+            // Pretend our store has a transaction each second that sells 4 hats
+            Thread.Sleep(1000);
+            s_hatsSold.Add(4);
+        }
+
+        meterListener.Dispose();
+    }
+
+    static void OnMeasurementRecorded<T>(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string,object>> tags, object? state)
```

Comment:
>T [](http://example.com/codeflow?start=38&length=1)

should we make this <int>?

Reply:
I was doing all my examples using 3.1 SDK where adding the `?` for nullability would create compiler warnings and I was attempting to avoid additional explanatory text/code to work around it. In many areas I know our new features are often coupled with users adopting new versions of the SDK/runtime but given our broad back-compat on DiagnosticSource I am guessing that the majority of users for this API are not yet using C# nullability analysis.

If you think it is likely to cause confusion we could still add some text about it though. Let me know your thoughts.

Reply:
I realized the `object? state` was already causing a nullability warning a missed somehow so I removed it for now to be consistent on not provided nullability markup.

Reply:
Oh, my comment was not specific to nullability. I see my comment was malformed. sorry about that. My comment is to use `int` instead of `T` on this callback method? just to educate users using specific type would have a better perf. But I am ok too if you think using `T` is simpler for this example.

Reply:
Ah no worries Tarek. I'll stick with the T for now because I think it will let users extend the scenario more easily. I haven't measured, but as best I understand this isn't a case where the generics are creating perf overhead.